README
NAME

    Devel::Examine::Subs - Get info about, search/replace and inject code
    into Perl files and subs.

SYNOPSIS

        use Devel::Examine::Subs;
    
        my $file = 'perl.pl'; # or directory, or Module::Name
        my $search = 'string';
    
        my $des = Devel::Examine::Subs->new( file => $file );

    Get all sub names in a file

        my $aref = $des->all;

    Get all the subs as objects

        $subs = $des->objects;
    
        for my $sub (@$subs){
            $sub->name;       # name of sub
            $sub->start;      # first line number of sub in file
            $sub->end;        # last line number of sub in file
            $sub->line_count; # number of lines in sub
            $sub->code;       # entire sub code from file
            $sub->lines;      # lines that match search term
        }

    Get the sub objects within a hash

        my $subs = $des->objects( objects_in_hash => 1 );
    
        for my $sub_name (keys %$subs) {
    
            print "$sub_name\n";
    
            my $sub = $subs->{$sub_name};
    
            print $sub->start . "\n" .
                  $sub->end . "\n";
                  ...
        }

    Get all subs containing "string" in the body

        my $aref = $des->has( search => $search );

    Search and replace code in subs

        $des->search_replace( exec => sub { $_[0] =~ s/this/that/g; } );

    Inject code into sub after a search term (preserves previous line's
    indenting)

        my @code = <DATA>;
    
        $des->inject_after(
                        search => 'this',
                        code => \@code,
                      );
    
        __DATA__
    
        # previously uncaught issue
    
        if ($foo eq "bar"){
            confess 'big bad error';
        }

    Print out all lines in all subs that contain a search term

        my $subs = $des->lines(search => 'this');
    
        for my $sub (keys %$subs){
    
            print "\nSub: $sub\n";
    
            for my $line (@{ $subs->{$sub} }){
                my ($line_num, $text) = each %$line;
                say "Line num: $line_num";
                say "Code: $text\n";
            }
        }

    The structures look a bit differently when 'file' is a directory. You
    need to add one more layer of extraction.

        my $files = $des->objects;
    
        for my $file (keys %$files){
            for my $sub (@{$files->{$file}}){
                ...
            }
        }

    Print all subs within each Perl file under a directory

        my $files = $des->all( file => 'lib/Devel/Examine' );
    
        for my $file (keys %$files){
            print "$file\n";
            print join('\t', @{$files->{$file}});
        }

    Most methods can include or exclude specific subs

        my $has = $des->has( include => ['dump', 'private'] );
    
        my $missing = $des->missing( exclude => ['this', 'that'] );
    
        # note that 'exclude' param renders 'include' invalid

DESCRIPTION

    Gather information about subroutines in Perl files (and in-memory
    modules), with the ability to search/replace code, inject new code, get
    line counts, get start and end line numbers, access the sub's code and
    a myriad of other options.

METHODS

    See the PARAMETERS for the full list of params, and which ones are
    persistent across runs using the same object.

 new

    Mandatory parameters: file => $filename

    Instantiates a new object. If $filename is a directory, we'll iterate
    through it finding all Perl files. If $filename is a module name (eg:
    Data::Dump), we'll attempt to load the module, extract the file for the
    module, and load the file. CAUTION: this will be a production %INC file
    so be careful.

    Only specific params are guaranteed to stay persistent throughout a run
    on the same object, and are best set in new(). These parameters are
    file, extensions, maxdepth, cache, regex, copy, no_indent, backup and
    diff.

    Note: omit the file parameter if all you'll be using is the module()
    method.

 all

    Mandatory parameters: None

    Returns an array reference containing the names of all subroutines
    found in the file, listed in the order they appear in the file.

 has

    Mandatory parameters: search => 'term'

    Returns an array reference containing the names of the subs where the
    subroutine contains the search text.

 missing

    Mandatory parameters: search => 'term'

    The exact opposite of has.

 lines

    Mandatory parameters: search => 'text'

    Gathers together all line text and line number of all subs where the
    subroutine contains lines matching the search term.

    Returns a hash reference with the subroutine name as the key, the value
    being an array reference which contains a hash reference in the format
    line_number => line_text.

 objects

    Mandatory parameters: None

    Optional parameters: objects_in_hash => 1

    Returns an array reference of subroutine objects. If the optional
    objects_in_hash is sent in with a true value, the objects will be
    returned in a hash reference where the key is the sub's name, and the
    value is the sub object.

    See SYNOPSIS for the structure of each object.

 module('Module::Name')

    Mandatory parameters: 'Module::Name'. Note that this is one public
    method that takes its parameter in string format (as opposed to hash
    format).

    Note that this method pulls the subroutine names from the namespace
    (which may include includeed subs. If you only want a list of subs
    within the actual module file, send the module name as the value to the
    file parameter, and use the common methods (all, has, missing etc) to
    extract the names.

    Returns an array reference containing the names of all subs found in
    the module's namespace symbol table.

 order

    After one of the other user methods are called, call this method to get
    returned to you an array of the names of subs you collected, in the
    order that they appear in the file. By default, because we use hashes
    internally, subs aren't ever in proper order.

 search_replace

    Mandatory parameters: exec => $cref

    Core optional parameter: copy => 'filename.txt'

    Coderef should be created in the form sub { $_[0] =~ s/search/replace/;
    };. This allows us to avoid string eval, and allows us to use any regex
    modifiers you choose. The $_[0] element represents each line in the
    file, as we loop over them.

 replace

    Parameters: exec => $cref, limit => 1

    This is the entire file brother to the sub-only search_replace(). The
    limit parameter specifies how many successful replacements to do,
    starting at the top of the file. Set to a negative integer for
    unlimited (this is the default).

    The exec parameter is a code reference, eg: my $cref = sub {$_[0] =~
    s/this/that/;}. All standard Perl regular expressions apply, along with
    their modifiers. The $_[0] element represents each line in the file, as
    we loop over them.

    Returns the number of lines changed in file mode, and an empty hashref
    in directory mode.

 inject_after

    Mandatory parameters: search => 'this', code => \@code

    Injects the code in @code into the sub within the file, where the sub
    contains the search term. The same indentation level of the line that
    contains the search term is used for any new code injected. Set
    no_indent parameter to a true value to disable this feature.

    By default, an injection only happens after the first time a search
    term is found. Use the injects parameter (see PARAMETERS) to change
    this behaviour. Setting to a positive integer beyond 1 will inject
    after that many finds. Set to a negative integer will inject after all
    finds.

    The code array should contain one line of code (or blank line) per each
    element. (See SYNOPSIS for an example). The code is not manipulated
    prior to injection, it is inserted exactly as typed. Best to use a
    heredoc, __DATA__ section or an external text file for the code.

    Optional parameters:

    copy

      See search_replace() for a description of how this parameter is used.

    injects

      How many injections do you want to do per sub? See PARAMETERS for
      more details.

 inject

    Parameters (all are mutually exclusive, use only one):

    line_num => 33 with code => \@code or, inject_use => ['use
    Module::Name', 'use Module2::Name'] or, inject_after_sub_def => ['code
    line 1;', 'code line 2;']

    line_num will inject the block of code in the array reference
    immediately after the line number specified.

    inject_use will inject the statements prior to all existing use
    statements that already exist in the file(s). If none are found, will
    inject the statements right after a Package statement if found.

    Technically, you don't have to inject a use statement, but I'd advise
    it.

    inject_after_sub_def will inject the lines of code within the array
    reference value immediately following all sub definitions in a file.
    Next line indenting is used, and sub definitions with their opening
    brace on a separate line than the definition itself is caught.

 remove

    Parameters: delete => ['string1', 'string2']

    Deletes from the file(s) the entire lines that contain the search
    terms.

    This method is file based... the work happens prior to digging up subs,
    hence exclude, include and other sub-based parameters have no effect.

 backup(Bool)

    Configure whether to make a filename.bak copy of all files read by DES.
    A true value sent in will enable this feature, a false value will
    disable it. Returns 1 (true) if this feature is enabled, and 0 (false)
    if not.

    Disabled by default.

DEVELOPER METHODS

 valid_params

    Returns a hash where the keys are valid parameter names, and the value
    is a bool where if true, the parameter is persistent (remains between
    calls on the same object) and if false, the param is transient, and
    will be made undef after each method call finishes.

 run

    Parameter format: Hash reference

    All public methods call this method internally. This is the only public
    method that takes its parameters as a single hash reference. The public
    methods set certain variables (filters, engines etc). You can get the
    same effect programatically by using run(). Here's an example that
    performs the same operation as the has() public method:

        my $params = {
                search => 'text',
                post_proc => 'file_lines_contain',
                engine => 'has',
        };
    
        my $return = $des->run($params);

    This allows for very fine-grained interaction with the application, and
    makes it easy to write new engines and for testing.

 add_functionality

    WARNING!: This method is for development of this distribution only!

    While writing new processors, set the processor type to a callback
    within the local working file. When the code performs the actions you
    want it to, put a comment line before the code with #<des> and a line
    following the code with #</des>. DES will slurp in all of that code
    live-time, inject it into the specified processor, and configure it for
    use. See examples/write_new_engine.pl for an example of creating a new
    'engine' processor.

    Returns 1 on success.

    Parameters:

    add_functionality

      Informs the system which type of processor to inject and configure.
      Permitted values are 'pre_proc', 'post_proc' and 'engine'.

    add_functionality_prod

      Set to a true value, will update the code in the actual installed
      Perl module file, instead of a local copy.

    Optional parameters:

    copy

      Set it to a new file name which will be a copy of the specified file,
      and only change the copy. Useful for verifying the changes took
      properly.

 pre_procs, post_procs, engines

    For development. Returns the list of the respective built-in callbacks.

PARAMETERS

    There are various parameters that can be used to change the behaviour
    of the application. Some are persistent across calls, and others
    aren't. You can change or null any/all parameters in any call, but some
    should be set in the new() method (set it and forget it).

    The following list are persistent parameters, which need to be manually
    changed or nulled. Consider setting these in new().

    file

      State: Persistent

      Default: None

      The name of a file, directory or module name. Will convert module
      name to a file name if the module is installed on the system. It'll
      require the module temporarily and then 'un'-require it immediately
      after use.

      If set in new(), you can omit it from all subsequent method calls
      until you want it changed. Once changed in a call, the updated value
      will remain persistent until changed again.

    backup

      State: Persistent

      Default: Disabled

      Set this to a true value to have a .bak file copy created on all file
      reads. The .bak file will be created in the directory the script is
      run in.

    extensions

      State: Persistent

      Default: ['*.pm', '*.pl')]

      By default, we load only *.pm and *.pl files. Use this parameter to
      load different files. Only useful when a directory is passed in as
      opposed to a file. This parameter is persistent until manually reset
      and should be set in new().

      Values: Array reference where each element is the names of files to
      find. Any wildcard or regex that are valid in File::Find::Rule's
      <http://search.cpan.org/~rclamp/File-Find-Rule-0.33/lib/File/Find/Rule.pm>
      are valid here. For example, [qw(*.pm *.pl)] is the default.

    maxdepth

      When running in directory mode, how many levels deep do you want to
      traverse? Default is unlimited. Set to a positive integer to set.

    cache

      State: Persistent

      Default: Undefined

      If multiple calls on the same object are made, caching will save the
      file/directory/sub information, saving tremendous work for subsequent
      calls. This is dependant on certain parameters not changing between
      calls.

      Set to a true value (1) to enable. Best to call in the new method.

    copy

      State: Persistent

      Default: None

      For methods that write to files, you can optionally work on a copy
      that you specify in order to review the changes before modifying a
      production file.

      Set this parameter to the name of an output file. The original file
      will be copied to this name, and we'll work on this copy.

    regex

      State: Persistent

      Default: Enabled

      Set to a true value, all values in the 'search' parameter become
      regexes. For example with regex on, /thi?s/ will match "this", but
      without regex, it won't. Without 'regex' enabled, all characters that
      perl treats as special must be escaped. This parameter is persistent;
      it remains until reset manually.

    no_indent

      State: Persistent

      Default: Disabled

      In the processes that write new code to files, the indentation level
      of the line the search term was found on is used for inserting the
      new code by default. Set this parameter to a true value to disable
      this feature and set the new code at the beginning column of the
      file.

    diff

      State: Persistent

      Not yet implemented.

      Compiles a diff after each edit using the methods that edit files.

    The following parameters are not persistent, ie. they get reset before
    entering the next call on the DES object. They must be passed in to
    each subsequent call if the effect is still desired.

    include

      State: Transient

      Default: None

      An array reference containing the names of subs to include. This (and
      exclude) tell the Processor phase to generate only these subs,
      significantly reducing the work that needs to be done in subsequent
      method calls.

    exclude

      State: Transient

      Default: None

      An array reference of the names of subs to exclude. See include for
      further details.

      Note that exclude renders include useless.

    injects

      State: Transient

      Default: 1

      Informs inject_after() how many injections to perform. For instance,
      if a search term is found five times in a sub, how many of those do
      you want to inject the code after?

      Default is 1. Set to a higher value to achieve more injects. Set to a
      negative integer to inject after all.

    pre_proc_dump, post_proc_dump, engine_dump, cache_dump, core_dump

      State: Transient

      Default: Disabled

      Set to 1 to activate, exit()s after completion.

      Print to STDOUT using Data::Dumper the structure of the data
      following the respective phase. The core_dump will print the state of
      the data, as well as the current state of the entire DES object.

      NOTE: The 'post_proc' phase is run in such a way that the filters can
      be daisy-chained. Due to this reason, the value of post_proc_dump
      works a little differently. For example:

          post_proc => ['one', 'two'];

      ...will execute filter 'one' first, then filter 'two' with the data
      that came out of filter 'one'. Simply set the value to the number
      that coincides with the location of the filter. For instance,
      post_proc_dump => 2; will dump the output from the second filter and
      likewise, 1 will dump after the first.

      For cache_dump, if it is set to one, it'll dump cache but the
      application will continue. Set this parameter to an integer larger
      than one to have the application exit immediately after dumping the
      cache to STDOUT.

    pre_proc_return, post_proc_return, engine_return

      State: Transient

      Default: Disabled

      Returns the structure of data immediately after being processed by
      the respective phase. Useful for writing new 'phases'. (See "SEE
      ALSO" for details).

      NOTE: post_proc_return does not behave like post_proc_dump. It will
      only return after all post_procs have executed.

    config_dump

      State: Transient

      Default: Disabled

      Prints to STDOUT with Data::Dumper the current state of all loaded
      configuration parameters.

    pre_proc, post_proc, engine

      State: Transient

      Default: undef

      These are mainly used to set up the public methods with the proper
      callbacks used by the run() command.

      engine and pre_proc take either a single string that contains a valid
      built-in callback, or a single code reference of a custom callback.

      post_proc works a lot differently. These modules can be
      daisy-chained. Like engine and pre_proc, you can send in a string or
      cref, or to chain, send in an aref where each element is either a
      string or cref. The filters will be executed based on their order in
      the array reference.

REPOSITORY

    https://github.com/stevieb9/devel-examine-subs

BUILD REPORTS

    CPAN Testers: http://matrix.cpantesters.org/?dist=Devel-Examine-Subs

DEBUGGING

    If Devel::Trace::Subs is installed, you can configure stack tracing.

    In your calling script, set $ENV{DES_TRACE} = 1.

    See perldoc Devel::Trace::Subs for information on how to access the
    traces.

SEE ALSO

    perldoc Devel::Examine::Subs::Preprocessor

      Information related to the 'pre_proc' phase core modules.

    perldoc Devel::Examine::Subs::Postprocessor

      Information related to the 'post_proc' phase core modules.

    perldoc Devel::Examine::Subs::Engine

      Information related to the 'engine' phase core modules.

AUTHOR

    Steve Bertrand, <steveb at cpan.org>

SUPPORT

    You can find documentation for this module with the perldoc command.

        perldoc Devel::Examine::Subs

LICENSE AND COPYRIGHT

    Copyright 2016 Steve Bertrand.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Devel::Examine::Subs

1.75 UNREL
    -

1.74    2021-10-11
    - Added CI and coverage badges

1.73    2021-10-11
    - Add heading to forgotten example in SYNOPSIS
    - Re-layout Changes file

1.72    2021-10-11
    - Migrated from Travis CI to Github Actions
    - Added section headers for POD in SYNOPSIS

1.71    2020-12-04
    - fixes #31; unlink temporary file in _read_file() after we're done
      with it. Without unlink, we were leaking temporary files
    - add rw(), provides an interface to the internal File::Edit::Portable
      object
    - add tempfile(), provides an interface to the internal File::Temp
      object
    - added $ENV{ISSUE_31_TEST}, and modified _read_file() to allow tests
      to skip the recsep checks for testing against issue #31
    - updated file count tests to properly deal with the new test files
    - added t/58-tempfile_unlink_bug.t, which tests against the problem
      identified in issue #31
    - in POD, move description section above synopsis (closes #30)
    - fix misleading/incorrect comment (fixes #28)

1.70    2018-10-24
    - clean up code in respect to specific layouts (closes #18)
    - efficiency change in the core _proc() sub. We now check for existence
      before comparison (fixes #20)
    - cleaned up all instances of eval() (closes #22)
    - apply Pull Request that disables the win32 line-ending tests (t/41)
      that caused the test suite to hang on some non-Win32 platforms
      (thanks Slaven Rezić for the report, and thanks to Steve Dondley for
      the PR) (fixes #25)

1.69    2017-02-03
    - no operational/functional changes
    - forgot to add the t/57-heredoc-bug.t test file to MANIFEST, and this
      caused off-by-one test failures on Testers

1.68    2017-02-01
    - merge pull request #13, changes PPIs "content" with "serialize", as
      the former can mangle heredocs (thanks Varadinksy)
    - migrated the test files accompanying the PR into the normal test suite
    - added check in _read_file() to see if the file is a proper file
    - merged in PR #14 and PR #15 with some minor changes. Both PRs dealt
      with POD fixes and clarifications (thanks Anthony Brummett (brummett))

1.67    2016-03-14
    - bump version of File::Edit::Portable prereq to v1.24 due to bug fixes
      in previous versions of that module (File::Edit::Portable #35)
    - corrected logic in t/05-object.t tests and verified with accuracy
      that all tests are correct and working (fixes #11)
    - fixed test in t/21-core_dump.t to accommodate FEP updates

1.66    2016-02-23
    - order() dies if it's called on a directory structure, as opposed to a
      single file (closes #9)
    - changed all croak()s to confess() for better diagnostic output
    - POD fixes (closes #8)
    - bumped prereq of Mock::Sub to 1.06
    - bumped prereq of File::Edit::Portable to 1.19

1.65    2016-02-07
    - bumped prereq ver of File::Edit::Portable to 1.18

1.64    2016-02-04
    - removed erroneous build requirement that negated the File::Edit::Portable
      prereq_pm version update, allowing tests to continue to fail on some
      versions of FreeBSD

1.63    2016-02-03
    - added backup(), set/get status of _read_file() creating a '.bak' copy
      of each file on every read. Defaults to disabled (0) (fixes #7)
    - _read_file() now compares the file's line endings with that of the
      platform, and skips creating a tempfile if they match (performance).
      (fixes #4)
    - re-implemented t/002-critic.t, if RELEASE_TESTING is enabled, and
      Perl::Critic is installed
    - cosmetic changes to tests and test data files to satisfy critic
    - module() now checks to see if a module is already loaded, and then
      unloads it if it wasn't (fixes #5)


1.62    2016-01-26
    - we now require File::Edit::Portable v1.17, due to an issue in the
      previous version where an flock() problem in FBSD 10.1 and 9.2 amd64
      was causing tests to fail on those platforms

1.61    2016-01-18
    - _write_file() was breaking with file is a dir param and no copy param
      was sent in (closes #3)

1.60    2016-01-16
    - no functional changes
    - POD updates

1.59    2016-01-14
    - 100% test coverage!
    - when file param is a directory, we now use the copy param as a dir,
      create that dir, and copy all the files in the file param to it, and
      operate on those (in operations that write to files)
    - add_functionality() now checks to see if the sub being added already
      exists (closes #2)
    - true value at end of module remains in the correct place (closes #1)

1.58    20151230
    - _proc() now sets $p->{order}, which keeps track of the original order
      of subs in the file being worked on
    - added order(), returns the list of sub names in the order they appear
      in the file
    - fixed minor issues in a couple of unit test files (file cleanup on
      windows wasn't working correctly)
    - removed 'blead' from travis
    - win32 tests in CPAN Testers should be ok again after last dev version
    - re-enable t/24-preproc (seems to be working on win32 again)

1.57_01   20151120
    - certain deprications will be necessary after the rest of the module
      is rewritten to use PPI where it currently isn't
    - reworked _read_file(), it now populates $self->{params}{file_contents}
      from a PPI doc
    - _proc() modified to use PPI to get the sub code from the file,
      instead of extracting it out manually out of an array slice by
      counting line nums
    - $subs struct no longer has a copy of the entire file contents in each
      sub. This reduces the size of the data and all objects significantly

1.56    20151029
    - no noticeable functional changes
    - most examples have been removed as unnecessary
    - extra check put in before we clobber *trace() in BEGIN

1.55    20151022
    - no functional changes
    - runs down to perl v5.8

Makefile.PL
use 5.008;
use warnings;
use strict;

use ExtUtils::MakeMaker;

WriteMakefile
(
  PL_FILES => {},
  EXE_FILES => [],
  NAME => 'Devel::Examine::Subs',
  MIN_PERL_VERSION => '5.008',
  LICENSE => 'perl',
  INSTALLDIRS => 'site',
  VERSION_FROM => 'lib/Devel/Examine/Subs.pm',
  META_MERGE => {
        'meta-spec' => { version => 2 },
        resources   => {
             bugtracker => {
                web => 'https://github.com/stevieb9/devel-examine-subs/issues',
            },
            repository => {
                type => 'git',
                url => 'https://github.com/stevieb9/devel-examine-subs.git',
                web => 'https://github.com/stevieb9/devel-examine-subs',
            },
        },
    },
    CONFIGURE_REQUIRES => {
        'Cwd' => 0,
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Mock::Sub' => '1.06',
        'Test::More' => 0,
        'Test::Trap' => 0,
    },
    PREREQ_PM => {
       'Carp' => 0,
       'Data::Compare' => 0,
       'Data::Dumper' => 0,
       'File::Basename' => 0,
       'File::Edit::Portable' => '1.19',
       'PPI' => 0,
       'Symbol' => 0,
    },
);
